module/product/src/Infrastructure/Provider/FilteredQueryBuilderProvider.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+/**
+ * Copyright © Bold Brand Commerce Sp. z o.o. All rights reserved.
+ * See LICENSE.txt for license details.
+ */
+
+declare(strict_types=1);
+
+namespace Ergonode\Product\Infrastructure\Provider;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Query\QueryBuilder;
+use Ergonode\BatchAction\Domain\ValueObject\BatchActionFilter;
+use Ergonode\BatchAction\Infrastructure\Provider\FilteredQueryBuilderProviderInterface;
+use Ergonode\Core\Domain\Query\LanguageQueryInterface;
+use Ergonode\Grid\FilterGridConfiguration;
+use Ergonode\Product\Infrastructure\Factory\DataSet\DbalQueryBuilderProductDataSetFactory;
+use Ergonode\Product\Infrastructure\Grid\ProductGridBuilder;
+
+class FilteredQueryBuilderProvider implements FilteredQueryBuilderProviderInterface
+{
+    private DbalQueryBuilderProductDataSetFactory $productQueryBuilderFactory;
+
+    private ProductGridBuilder $gridBuilder;
+
+    private ProductIdsProvider $productIdsProvider;
+
+    private Connection $connection;
+
+    private LanguageQueryInterface $languageQuery;
+
+    public function __construct(
+        DbalQueryBuilderProductDataSetFactory $productQueryBuilderFactory,
+        ProductGridBuilder $gridBuilder,
+        ProductIdsProvider $productIdsProvider,
+        Connection $connection,
+        LanguageQueryInterface $languageQuery
+    ) {
+        $this->productQueryBuilderFactory = $productQueryBuilderFactory;
+        $this->gridBuilder = $gridBuilder;
+        $this->productIdsProvider = $productIdsProvider;
+        $this->connection = $connection;
+        $this->languageQuery = $languageQuery;
+    }
+
+
+    public function provide(?BatchActionFilter $filter): QueryBuilder
```

Comment:
Make it required and check that filter exist (or even $filter->getQuery()) move service that use this provider.
Also IMHO this is more builder then provider .